### ISSUE TITLE
coqide can't open

### ISSUE THREAD

**Issue by xxchan** [coq/coq#9799]
When I update to 8.9 (in Windows 10), CoqIDE can't open. Even if I uninstall and install again, it doesn't work.

Error message is:

Connection with coqtop failed!
Command was: ""C:\Coq\bin\coqidetop.opt.exe" -batch "
Answer was: While loading initial state:
  Warning: Cannot open C:\Coq\lib/theories [cannot-open-path,filesystem]
  While loading initial state:
  Warning: Cannot open C:\Coq\lib/plugins [cannot-open-path,filesystem]
  Error:
  Cannot find a physical path bound to logical path matching suffix
  <> and prefix Coq.
  
Exception was: Coq.WrongExitStatus("WEXITED 1")

**Comment by xxchan** [comment]
Then I tried Coq 8.8.2, and it worked normally, but when I installed 8.9.0 again, it failed again

**Comment by ejgallego** [comment]
Can you try with `master`? [There should be binaries somewhere]

**Comment by Zimmi48** [comment]
Windows binaries for the master branch are super hard to find if you don't give a pointer (the link is documented nowhere, for reasons).

https://gitlab.com/coq/coq/-/jobs/artifacts/master/file/artifacts/coq-8.10+alpha-installer-windows-x86_64.exe?job=windows64
and
https://gitlab.com/coq/coq/-/jobs/artifacts/master/file/artifacts/coq-8.10+alpha-installer-windows-i686.exe?job=windows32

@ejgallego Do you have any reason to believe that the problem could be a regression in 8.9 that has disappeared in master?

**Comment by ejgallego** [comment]
> Do you have any reason to believe that the problem could be a regression in 8.9 that has disappeared in master?

Yes I do; there have been a large amount of changes around that area, so I'd be good to test.

**Comment by xxchan** [comment]
@ejgallego I tried the version @Zimmi48 provided, and the same error occured again.

**Comment by jfehrle** [comment]
@xchan CoqIDE will start if you remove COQLIB from your environment variables.  CoqIDE is not constructing the paths correctly.  In your output above, the `/` in `C:\Coq\lib/theories` is incorrect; it should be a `\`.

**Comment by jfehrle** [comment]
This is a regression.  8.8.2 64 bit works with a variety of COQLIB values, e.g.

```
/cygdrive/c/Coq8.8.2-64/lib
/cygdrive/c/Coq8.8.2-64
C:\Coq8.8.2-64\lib
C:\Coq8.8.2-64
```

**Comment by SkySkimmer** [comment]
I think COQLIB was just not considered by coq pre 8.9

Gaëtan Gilbert

**Comment by gares** [comment]
Question is, who sets COQLIB?

**Comment by jfehrle** [comment]
> I think COQLIB was just not considered by coq pre 8.9

That appears to be the case since CoqIDE also comes up if COQLIB is "garbage".

**Comment by jfehrle** [comment]
> Question is, who sets COQLIB?

Should I read that as a neutral question or perhaps "who in their right mind sets COQLIB"?

If it's a useful feature, we should fix it promptly.  I think in 8.10.  It's an easy fix.  I can do it.

If you're aware of any other paths that may have the same issue, please let me know.

**Comment by Zimmi48** [comment]
>If it's a useful feature, we should fix it promptly. I think in 8.10.

That even sounds like something that should be backported.

**Comment by gares** [comment]
Maybe I misunderstood. Honouring the value of Coqlib seems useful to me. But it must be correctly set. So my question is: how is that we end up with a wrong value? Which scenario? It is not the installer that sets it, right?

**Comment by Zimmi48** [comment]
That's a question that I'm also interested in knowing the answer to. Especially as the OP did not confirm yet that unsetting this variable did actually solve their problem.

But note that "wrong value" is a bit strong in that case. There are plenty of Windows tools that support paths mixing `/` and `\` and it would be good if it was the case of Coq tools as well.

**Comment by jfehrle** [comment]
When I compile `master` with a simple `make` rather than the windows .bat file used by the official build, `coqtop` uses `\` to add to the path, which works fine if I pass it a Windows style pathname.

When I compile `remote/upstream/V8.9` the same way (which AFAIK should have source for the released version of 8.9), it uses '/' to add to the path.  I don't see how this for Windows.

When I run the official 8.9 downloaded from the Coq website, `coqtop` uses `/` to add to the path, which gives invalid Windows paths.

AFAIK Coq does not support cygwin-style pathnames.

Perhaps the .bat file and the make builds are not equivalent.  Or maybe I've gotten confused.
@MSoegtropIMC do you have any thoughts?

master build with make:
```
$ COQLIB=`pwd`

$ bin/coqtop.exe
Welcome to Coq 8.10+alpha (March 2019)
Warning:
Cannot open Raised by primitive operation at file "vernac/mltop.ml", line 181, characters 44-71
/cygdrive/c/Users/Jim.DESKTOP-MMPNVGH/Documents/coqSource/coq\theories
[cannot-open-path,filesystem]
Warning:
Cannot open Raised by primitive operation at file "vernac/mltop.ml", line 181, characters 44-71
/cygdrive/c/Users/Jim.DESKTOP-MMPNVGH/Documents/coqSource/coq\plugins
[cannot-open-path,filesystem]
Error: Cannot find a physical path bound to logical path matching suffix
<> and prefix Coq.

$ COQLIB=`cygpath -w $COQLIB`

$ bin/coqtop.exe
Welcome to Coq 8.10+alpha (March 2019)

Coq <
```

8.9 built with make:
```
$ COQLIB=`pwd`

$ bin/coqtop.exe
Welcome to Coq 8.9+alpha (March 2019)
While loading initial state:
Warning:
Cannot open /cygdrive/c/Users/Jim.DESKTOP-MMPNVGH/Documents/coqSource/coq/theories
[cannot-open-path,filesystem]
While loading initial state:
Warning:
Cannot open /cygdrive/c/Users/Jim.DESKTOP-MMPNVGH/Documents/coqSource/coq/plugins
[cannot-open-path,filesystem]
Error:
Cannot find a physical path bound to logical path matching suffix
<> and prefix Coq.

$ COQLIB=`cygpath -w $COQLIB`

$ bin/coqtop.exe
Welcome to Coq 8.9+alpha (March 2019)

Coq <
```

8.9 downloaded from the Coq website:

```
$ COQLIB=`pwd`

$ export COQLIB

$ bin/coqtop.exe
Welcome to Coq 8.9.0 (January 2019)
While loading initial state:
Warning: Cannot open /cygdrive/c/Coq8.9-64/theories
[cannot-open-path,filesystem]
While loading initial state:
Warning: Cannot open /cygdrive/c/Coq8.9-64/plugins
[cannot-open-path,filesystem]
Error:
Cannot find a physical path bound to logical path matching suffix
<> and prefix Coq.

$ COQLIB=`cygpath -w $COQLIB`

$ bin/coqtop.exe
Welcome to Coq 8.9.0 (January 2019)
While loading initial state:
Warning: Cannot open C:\Coq8.9-64/theories [cannot-open-path,filesystem]
While loading initial state:
Warning: Cannot open C:\Coq8.9-64/plugins [cannot-open-path,filesystem]
Error:
Cannot find a physical path bound to logical path matching suffix
<> and prefix Coq.
```

**Comment by jfehrle** [comment]
> It is not the installer that sets it, right?

No.  I do the install outside of cygwin.  I have no Windows environment COQLIB variable set.  Nor does my Cygwin start up with such a value.  So it would seem the user set this.

> There are plenty of Windows tools that support paths mixing / and \ and it would be good if it was the case of Coq tools as well.

We would also need to translate "/cygdrive/c/..." to "C:\..." (simple) and for "/foo" we need to insert the Windows format home directory of Cygwin as a prefix (probably also simple by running `cygpath -w XXX` and catching any "missing program" error).  But finding all the right places to change the code may take some doing.  My initial attempt wouldn't build.  It produced many lines of these messaages:

```
File "topbin/coqc_bin.ml", line 1:
Error: No implementations provided for the following modules:
         Typeops referenced from library/library.cmxa(Global),
           interp/interp.cmxa(Notation), interp/interp.cmxa(Impargs),
           printing/printing.cmxa(Prettyp),
           tactics/tactics.cmxa(Elimschemes), tactics/tactics.cmxa(Hints),
           vernac/vernac.cmxa(Class), vernac/vernac.cmxa(Auto_ind_decl),
           vernac/vernac.cmxa(Search), vernac/vernac.cmxa(Indschemes),
           vernac/vernac.cmxa(Classes), vernac/vernac.cmxa(Record),
```

**Comment by MSoegtropIMC** [comment]
Since these are MinGW applications they cannot handle cygwin paths, so if you say
```
COQLIB=`pwd`
```
in cygwin it cannot possibly work. This should work
```
COQLIB=`cygpath -w <correct library path>`
```
In your case it doesn't work because the provided path is wrong. If I ask coqtop for the correct path I get:
```
>C:\bin\Coq-v8.9\bin\coqtop -where
C:\bin\Coq-v8.9/lib/coq
```
So I guess it will work for you if you append lib/coq to your path. Btw.: the coq library path changed some time back (around 8.7 or so). It should also work if you don't define COQLIB at all (I never do). If you don't define it, the path is constructed from the path of the executable.

**Comment by MSoegtropIMC** [comment]
Refering to the original error report: I would expect that it does work if you simply remove all definitions of COQLIB from your system. I use 8.9 regularly and it definitely does work. The Windows installer is made such that the library path is derived from the executable path. You can install any number of Coq installations in parallel and each will find its proper library. But obviously this can only work if COQLIB is not defined by the system.

**Comment by MSoegtropIMC** [comment]
P.S.: I just did a test and renamed the installation folder of an 8.9 installation. The effect is that coqtop -where returns the modified path. This shows that the library path is indeed derived from the path of the executable. I made sure to call coqtop from somewhere else to ensure that it has nothing to do with the current path.
So in summary, please don't define COQLIB, and if you think you must define it for whatever reason, define it in the correct way (run coqtop -where to see the correct path).

@jfehrle : please note that this is not true for a cygwin coq compiled with configure and make. If coq is build in this way, paths should be taken from the paths given during configure.

**Comment by jfehrle** [comment]
> In your case it doesn't work because the provided path is wrong. If I ask coqtop for the correct path I get:

```
>C:\bin\Coq-v8.9\bin\coqtop -where
C:\bin\Coq-v8.9/lib/coq
```
I see you're getting back a path with a mix of `\` and `/`.

I get something different:

local build of 8.9:
```
Jim@DESKTOP-MMPNVGH /cygdrive/c/Users/Jim.DESKTOP-MMPNVGH/Documents/coqSource/coq
$  bin/coqtop.exe -where
C:\Users\Jim.DESKTOP-MMPNVGH\Documents\coqSource\coq
```

official build of 8.9:

```
Jim@DESKTOP-MMPNVGH /cygdrive/c/Coq8.9-64
$ bin/coqtop.exe -where
C:\Coq8.9-64
```

Also, I can't use backslashes and `C:` in the command line the way you did.  Maybe you're not using `bash`?

```
$ C:\Coq8.9-64\bin\coqtop -where
bash: C:Coq8.9-64bincoqtop: command not found
```

**Comment by MSoegtropIMC** [comment]
@jfehrle : 
```
$ bin/coqtop.exe -where
C:\Coq8.9-64
```
Are you sure you don't have COQLIB defined to this path? I don't see how coqtop of the official installer could possibly return something not ending in lib/coq unless COQLIB is defined.


> Also, I can't use backslashes and `C:` in the command line the way you did. Maybe you're not using `bash`?

No this was all from a Windows command prompt. If you use a posix system, backslashes must be escaped. And if you supply any path to a MinGW coqtop from cygwin, you must make sure that they are in Windows format (it may include / but may not include /cygdrive).

**Comment by jfehrle** [comment]
> Are you sure you don't have COQLIB defined to this path?

Yeah, my bad.  Clearing COQLIB I get paths ending in `lib/coq`.

```
Jim@DESKTOP-MMPNVGH /cygdrive/c
$ /cygdrive/c/Coq8.9-64/bin/coqtop.exe -where
C:\Coq8.9-64/lib/coq
Jim@DESKTOP-MMPNVGH /cygdrive/c
$ echo $COQLIB
C:\Coq8.9-64\lib\coq

Jim@DESKTOP-MMPNVGH /cygdrive/c
$ /cygdrive/c/Coq8.9-64/bin/coqtop.exe
Welcome to Coq 8.9.0 (January 2019)

Coq < Quit.
```

Also I can start the 8.9 release verison of CoqIDE now.  Not sure how the OP got the error:

```
Jim@DESKTOP-MMPNVGH /cygdrive/c/Coq8.9-64
$ bin/coqtop -where
C:\Coq8.9-64/lib/coq

Jim@DESKTOP-MMPNVGH /cygdrive/c/Coq8.9-64
$ COQLIB="C:\Coq8.9-64\lib\coq"

Jim@DESKTOP-MMPNVGH /cygdrive/c/Coq8.9-64
$ export COQLIB

Jim@DESKTOP-MMPNVGH /cygdrive/c/Coq8.9-64
$ bin/coqide.exe

Jim@DESKTOP-MMPNVGH /cygdrive/c/Coq8.9-64
$ cd /

Jim@DESKTOP-MMPNVGH /
$ /cygdrive/c/Coq8.9-64/bin/coqtop.exe
Welcome to Coq 8.9.0 (January 2019)

Coq < Quit.

Jim@DESKTOP-MMPNVGH /
$ /cygdrive/c/Coq8.9-64/bin/coqide.exe

Jim@DESKTOP-MMPNVGH /
$
```

**Comment by MSoegtropIMC** [comment]
@jfehrle :

> Jim@DESKTOP-MMPNVGH /cygdrive/c/Coq8.9-64
> $ COQLIB="C:\Coq8.9-64\lib\coq"

There is (or should be) no need to define COQLIB in order to run coqide or coqtop from the official Windows installer. My recommended practice is to not define it. I cannot imagine a situation where this would do any good except if you want to experiment with different variants of the library.

@xxchan : can you please confirm that it works if you clear all definitions of COQLIB (in the Windows environment and possibly in shell scripts you use to start coqide).

**Comment by xxchan** [comment]
> @xxchan : can you please confirm that it works if you clear all definitions of COQLIB (in the Windows environment and possibly in shell scripts you use to start coqide).

@MSoegtropIMC : Oh, I can start coqide now after I deleted COQLIB in the environment and installed 8.9

**Comment by MSoegtropIMC** [comment]
@xxchan : thanks for confirming. I will close the issue then.

**Comment by gares** [comment]
We may still provide a better error message (that includes how Coq decided to look for the library in a place where it can't find it...).

BTW @xxchan can you tell us how you ended up setting COQLIB? Did you follow a tutorial somewhere? Was it suggested by another error message?

**Comment by MSoegtropIMC** [comment]
> We may still provide a better error message

Yes, indicating the possible sources for COQLIB and which one was picked in case this fails would make sense.

Another / additional option might be to give a warning in the installer if COQLIB is set.

Problems with this have been fairly rare so far. As far as I know this behavior didn't change since I made the scripts.

@SkySkimmer : you wrote above that Coq prior 8.9 possibly ignored COQLIB. Was this an assumption from the symptoms or do you know of such a change?

**Comment by xxchan** [comment]
> BTW @xxchan can you tell us how you ended up setting COQLIB? Did you follow a tutorial somewhere? Was it suggested by another error message?

@gares : I did nothing other than simply remove COQLIB from the environment variables... I compiled my earlier project and didn't find anything wrong till now.

**Comment by SkySkimmer** [comment]
See 200fbd0cec4709bd1fb5f7a32e1c40270e343b5a

Gaëtan Gilbert

**Comment by MSoegtropIMC** [comment]
@SkySkimmer : thanks, interesting!

Since honoring COQLIB is a new feature in 8.9, I think it is worth discussing to have a configure option to ignore it. As I said, with the executables in the Windows installer it hardly ever makes sense to use the value from COQLIB.

**Comment by MSoegtropIMC** [comment]
@xxchan : what @gares (and me) wanted to know is how it happened that COQLIB is set on your system. The installer doesn't set it. So I would think you have set it manually. The question is: why did you set it manually? Is there some tutorial suggesting to do this? If yo, could you point us to it, so that we can have it fixed?

**Comment by Zimmi48** [comment]
> As I said, with the executables in the Windows installer it hardly ever makes sense to use the value from COQLIB.

Disabling the feature at configure time looks a bit strong, especially as the user is not normally supposed to set this environment variable, so if they do (provided that it's not a tutorial that tells them to do it), then they are supposed to know why they do it.

**Comment by MSoegtropIMC** [comment]
> Disabling the feature at configure time looks a bit strong, especially as the user is not normally supposed to set this environment variable, so if they do (provided that it's not a tutorial that tells them to do it), then they are supposed to know why they do it.

I don't think that many users who have a reason to set COQLIB (for other reasons than just pointing Coq to the one and only correct coq library, which is not required on Windows) work with the Windows binaries, but for sure we can wait how this develops before we take any action.

**Comment by SkySkimmer** [comment]
In envars https://github.com/vbgl/coq/blob/11d1c70ce0747d03c849df1362119d4ccaccb443/lib/envars.ml#L97-L106 we verify that the coq_config.coqlib does contain the prelude.vo. We could do the same for COQLIB, so that if it's set to something invalid we use coq_config and warn.
ie
- unset COQLIB -> current behaviour
- valid COQLIB -> use it
- invalid COQLIB -> as unset COQLIB, except warn with "invalid COQLIB=blabla: blabla/theories/Prelude.vo does not exist"

**Comment by xxchan** [comment]
@MSoegtropIMC : Oh, sorry... 
In my memory I didn't set it manually...
But I just tried to search and found some tutorials about "Coq + ProofGeneral + GNU Emacs + Windows" online, which exactly mentioned set COQLIB. 
Although I did't use ProofGeneral, I think it is possible that at the very beginning when I didn't compile coq correctly, I saw these tutorials and randomly setted COQLIB, but I'm not very sure.

**Comment by MSoegtropIMC** [comment]
> * unset COQLIB -> current behaviour
> * valid COQLIB -> use it
> * invalid COQLIB -> as unset COQLIB, except warn with "invalid COQLIB=blabla: blabla/theories/Prelude.vo does not exist"

It is still a bit dangerous. COQLIB might point to a path which is valid in the sense that Prelude.vo exists, but it might be from a different version of Coq. I guess one would get some interesting error messages then as well. From CoqIDE such error messages might not be visible.

Can we discuss what the valid use cases for setting COQLIB are in the Windows installer environment? If there really is a use for setting COQLIB on Windows, I would only take it if one also defines "COQLIB_I_KONW_WHAT_I_AM_DOING"

**Comment by Zimmi48** [comment]
> "COQLIB_I_KONW_WHAT_I_AM_DOING"

I don't see this improving the situation: if tutorials tell people to set `COQLIB`, they could tell people to set `COQLIB_WHATEVER` just as well.

**Comment by MSoegtropIMC** [comment]
> I don't see this improving the situation: if tutorials tell people to set `COQLIB`, they could tell people to set `COQLIB_WHATEVER` just as well.

Well people might remember it better then ;-)

But jokes aside: can someone tell me why one would like to set COQLIB to something else? Since the library version must match the Coq version exactly there is very little room for experimenting. If someone really absolutely needs a different COQLIB for some interesting experiments, (s)he can still pass it as command line argument (this also works with CoqIDE). I would think that this is much safer and more appropriate for such situations.

So my vote is really to ignore this environment variable in the executables build especially for the Windows installer.

**Comment by jfehrle** [comment]
I think @SkySkimmer has the right idea.  There should be a more specific check that COQLIB is valid and an associated error message that says "COQLIB is invalid" and maybe also "... and typically should not be set".  And make sure that message is visible in CoqIDE.

I noticed a couple details that are worth cleaning up at some point:

1. Coq is inconsistent in several places on the use of `/` and `\`, e.g.

```
$ bin/coqtop -where
C:\Coq8.9-64/lib/coq
```
This is not a valid value for COQLIB on Windows.

2. The inconsistency in the paths to `theories` and `plugins` between a local build and the official installer is confusing.  Would be nice to make these consistent; it can cause some confusion.  Also `coq` is the only directory in `lib`, seems like that's not needed.

**Comment by SkySkimmer** [comment]
>This is not a valid value for COQLIB on Windows.

How can it work if it's invalid?

**Comment by ejgallego** [comment]
It does not.

I thought I had fixed almost all that cases tho.

**Comment by jfehrle** [comment]
It worked in my experiments because I explicitly replaced the `/` with `\`.

**Comment by ejgallego** [comment]
Can you try this diff?
```diff
--- a/lib/envars.ml
+++ b/lib/envars.ml
@@ -21,7 +21,7 @@ let safe_getenv warning n =
   )
 
 let ( / ) a b =
-  if Filename.is_relative b then a ^ "/" ^ b else b
+  if Filename.is_relative b then a ^ Filename.dir_sep ^ b else b
 
 let coqify d = d / "coq"
 
```

**Comment by jfehrle** [comment]
I was just experimenting with that but surprisingly it is not needed for a local build.  The local build on master always correctly uses `/` when appending to the path even without the change.  And the `-where` shows what it should.

I finally installed the 8.10-alpha from CI that @Zimmi48 referenced.  With COQLIB unset, the `-where` incorrectly includes '/'.  Would be nice to understand what's different.  I'll submit a PR with `dir_sep` and install the result to see if that makes any difference.

```
Jim@DESKTOP-MMPNVGH /cygdrive/c/Coq8.10-alpha-64
$ bin/coqtop.exe -where
C:\Coq8.10-alpha-64/lib/coq
```

**Comment by SkySkimmer** [comment]
The logic when it finds a local coqlib is different, see guess_coqlib.

**Comment by jfehrle** [comment]
Commenting out the `not Coq_config.local` in guess_coqlib makes no difference.

Also looks like Sys.file_exists is agnostic on the separator, so `let prelude = "theories/Init/Prelude.vo"` is OK:

```
# Sys.file_exists "bin/coqtop.exe";;
- : bool = true
# Sys.file_exists "bin\\coqtop.exe";;
- : bool = true
```

**Comment by ejgallego** [comment]
@jfehrle the problem only shows on absolute paths, @MSoegtropIMC described it accurately IMO.

**Comment by jfehrle** [comment]
> the problem only shows on absolute paths

I don't understand.  Which problem?  My most recent example that shows a problem when COQLIB is unset.  (And BTW almost all my experiments have used absolute paths.)

**Comment by MSoegtropIMC** [comment]
@jfehrle 

> $ bin/coqtop -where
> C:\Coq8.9-64/lib/coq

As I already explained this is a perfectly valid Windows path. Every version of Windows since NT 3.5 understands such paths natively in the API. Try e.g. at a CMD prompt:

```
dir "C:/Program Files"
```

What of cause doesn't work is using \ in bash - I guess this is more likely your problem if you feel you have to fiddle around with the paths.

**Comment by jfehrle** [comment]
I wasn't aware Windows had started to allow forward slashes, but they aren't always accepted when backslashes are.  They must be quoted:

```
C:\Users\Jim.DESKTOP-MMPNVGH>dir c:/
Invalid switch - "".

C:\Users\Jim.DESKTOP-MMPNVGH> dir "c:/"
 Volume in drive C is Windows
 Volume Serial Number is 9C5E-F983

 Directory of c:\
       :

C:\Users\Jim.DESKTOP-MMPNVGH>dir c:/Users
Invalid switch - "Users".
```

**Comment by MSoegtropIMC** [comment]
> I wasn't aware Windows had started to allow forward slashes

as far as I know it has always been like this, even in DOS. Definitely all Windows NT derivatives support / as path separator (that is NT 3.5, NT 3.51, NT 4.0, XP, ...). The problem is more that / is treated specially in command line parsing by CMD. There is a DOS CMD config option "SWITCHAR=" which allows to change the switch character from / to -, but it was never required to enable / as path separator.

To come to the point: the underlying system APIs support / as path separator since ever. What might mess it up is a shell, but as you wrote quoting usually fixes it. The OCaml system command seems to take reasonably care of the quoting. I have never seen a case where a mixed \ / path fed to the OCaml system command resulted in trouble.

In my experience if you want to call OCaml based MinGW executables from cygwin, the best approach is to use Windows path with / instead of \. If you do this from bash, there is usually no need for quoting.

**Comment by MSoegtropIMC** [comment]
I would still vote for ignoring COQLIB in coqide as installed by the Windows installer, since I don't see any useful application for it, but it can do a lot of harm. But I have no objection for keeping support in all other Versions, say Linux or even Windows opam builds.
